### DISCUSSION TITLE
X85A smart grid / 4 level power limitation

### DISCUSSION THREAD

**Discussion by raomin** [raomin/ESPAltherma#170] [Announcements]
On my Altherma I have a 6pin port X85A that is supposed to be connected to a *demand pcb* that can be connected to a smart grid information device that would limit the power consumption to 4 level.

This is actually complementary to the SG features of Altherma.
An Altherma Heat pump can usually be connected to a 2 switch info on smart grid. See eg:
![image](https://user-images.githubusercontent.com/4460208/196805834-9f139e67-3b60-4b2a-9dd8-b9e612292acb.png)
Then depending on the status of the 2 switches, it defines different 3 operation mode 
![image](https://user-images.githubusercontent.com/4460208/196806404-dbbf0506-81d5-4ce3-9fa1-c51af6f21639.png)

Basically:

- Free running is normal,
- Force off, nothing just defrost/emergency stuff
- Recommended on: if you can heat up a bit more DHW and rooms if needed.
- Force on: consume everything you need!

Now this X85A port more looks like a 'load shedding' system to limit the maximum power of the entire house
(lighting, domestic appliances, space heating…).

![image](https://user-images.githubusercontent.com/4460208/196530458-5b19bcdd-212d-4bd0-ad37-4fedfa970d23.png)

I found some time to have a look at this port and I'm happy to report that we will soon be able to control the 4 level of power consumption through ESPAltherma!

Basically, pin 6 is GND, then four level pins all pulled-up:
- pin 5: L4
- pin 4: L3
- pin 3: L2
- pin 2: L1

When pulled down, it sets the power control to the limitation level.
If eg L2 and L4 are down, the highest takes precedence.

So we would just need more pins to connect to the 4 pins and put them to GND and as it's the same GND as ESPAltherma: no need for relays.
The M5Stick-c would not be enough unless we use [a GPIO extender](https://shop.mchobby.be/en/grove/1913-m5stack-pca9554pw-extension-8-entreesortie-grove-3232100019133-m5stack.html). I guess I'll switch to a t-display but I'll also implement support for the extender.

**Comment by stomko11** [comment]
Port X85A is on my board, but according to manual or control panel, I do not have any SmartGrid capabilities available. Inside unit EHBH11CB9W. Is this something I may be able to utilize?

**Reply by raomin** [reply]
From what I can read, this is what manuals refer to _Power consumption control_ you can set 4 level to specific Amp consumption: [see here](https://www.daikin.fr/content/dam/document-library/Installer-reference-guide/heat/air-to-water-heat-pump-low-temperature/erga-ev/ERGA-(EV_EVH_EVH7)_ERGA04-08EVA_EHB(H_X)-(6V_9W)_Installer%20reference%20guide_4PEN629085-1C_English.pdf#page=217&zoom=100,65,833)

normally you need _Demand PCB (option EKRP1AHTA)_ to use it or ESPAltherma 😉
So yes, you'll soon can ;)

**Reply by DieterB61** [reply]
Hi, I'm so interested in this feature (I have an Altherma 3 LT/2023). In my manual it is also listed that option EKRP1AHTA is necessary for the control of power limit. My plumber gave me an offer of 251€ for that part. Moreover I tried the X5M SG function **but the power was not limited** (I'm in contact with Daikin about that problem).
Now my question: did I understand it correct, that there must be only a bridge between pin 6 (ground) and any of the pins 2-5 to limit the power (when everything else is set up in the settings)? Thus can I test this function with a simple cable connection between pin 6 and one of the pins 2-5?
Will this feature be included in ESPAltherma in near future?

**Reply by javimurcia** [reply]
That optional pcb is doing exactly that, but using optocouplers for isolation.

You can do the test you are describing, and if it works, check out https://github.com/raomin/ESPAltherma/pull/262 which is implementing this functionality.

**Reply by DieterB61** [reply]
Hi, can it be that you mixed up PIN 1 and 6?
And what do you mean with "nothing happens"?
You first have to configure the power limitation featue (in my heat pump it is [9.9.1] that the power limitation feature is enabled.
And you have to set the limits for all 4 levels.
After that the power consumption should be limited when you do the connections in X85A to the appropriate level. 
But will only see the result when the heat pump is running and you are able to measure the power consumption.

**Reply by DieterB61** [reply]
OK, I'm affraid I can not help. It looks like my one but restarting after connecting 6 to any other (except 1) sounds bad.
On the other hand it seems unlikly that there is another pin assignment on your board?

Beside the setting of the levels, did you assign that the power consumtion limit is enabled?

When you write, the power consumption is not limited you are probably able to measure it. 
I do it with an external shelly. Where do you see your power consumption?

**Reply by gerione** [reply]
I was checking it under the menu if power limit is enabled and in the messages in ESPALTHERMA. 
![image](https://github.com/raomin/ESPAltherma/assets/59999828/02903976-5993-4937-8166-24a5bb2e1b1a)

I will check. Thanks for the feedback.

**Comment by qris** [comment]
I haven't checked physically, but the installation manual for the EHBH/X 04-16 (which I have) says that this connector is on the main board and used to connect the demand PCB.

**Comment by alexdelli** [comment]
I’ve got it!!!


![2070C29D-2115-44E4-9FA5-4B85EAD46AD9](https://user-images.githubusercontent.com/6674409/196539882-7c2955d0-54c6-4f88-8cc6-75454272f865.jpeg)

**Comment by raomin** [comment]
@andreapier at first glance it would offer more flexibility (and a global support) than the SG1/SG2 you implemented. But what's your opinion?

**Reply by Berndie01** [reply]
Hi Ronin, your idea sounds great. I use a ESP-32 Dev Kit C V4, does it have enough IO Pins for that? I would like to use this in addition to heat up my domestic hot water only if pv solar delivers more than I can actually consume. Do you think that would be possible?
Cheers Berndie

**Reply by raomin** [reply]
Not sure. From what I've seen it works more like this: You setup 4 level of energy consumption limit (in Amps) and whatever happens, the heat pump will not consume more than theses level.
It's more a way to moderate consumption than a way to trigger consumption...

**Reply by raomin** [reply]
Yes, so you need to set your heat pump with the SG feature ON and add 2 relay to control the consumption, that's what @andreapier as implemented in ESPAltherma.
You can find more info on setting up this here: https://www.daikin.fr/content/dam/document-library/Installer-reference-guide/heat/air-to-water-heat-pump-low-temperature/erga-ev/ERGA-(EV_EVH_EVH7)_ERGA04-08EVA_EHB(H_X)-(6V_9W)_Installer%20reference%20guide_4PEN629085-1C_English.pdf#page=131&zoom=100,65,876

**Reply by andreapier** [reply]
> @andreapier at first glance it would offer more flexibility (and a global support) than the SG1/SG2 you implemented. But what's your opinion?

First of all let me tell you that this is a **great** news, thanks for taking the time to study it! Removing the need for 2 relays would be good. Also, having a finer control of consumption allows to optimize home-power budgeting even better!
I agree with you that this could be more flexible, and also support more PCBs, but I'd like to ask some questions so we can decide if this should replace current SG implementation (of course, if you know the answers!):
- Can we map the 4 levels to a known level of consumption (ie: level 1 = 1kw, level 2 = 2kw)? This should of course be specific to the heat pump model, as consumption can vary greatly depending on the external unit power. Or is it kind of a black box?
- Do the 4 levels allow to achieve what is currently possible setting SG2 or SG3 (setting higher DHW and room temperature)?

**Reply by raomin** [reply]
- Can we map the 4 levels to a known level of consumption (ie: level 1 = 1kw, level 2 = 2kw)? This should of course be specific to the heat pump model, as consumption can vary greatly depending on the external unit power. Or is it kind of a black box?

Yes you can set limits in term of Amp or KWH. See doc I've linked.

- Do the 4 levels allow to achieve what is currently possible setting SG2 or SG3 (setting higher DHW and room temperature)?

No, that's why I now understand that it will not replace SG2/SG3 relay function. It's a complementary function to fine tune consumption but not changing HP behavior.

**Reply by andreapier** [reply]
> Yes you can set limits in term of Amp or KWH. See doc I've linked.

Sorry, I didn't read it yet, but I should have. Will read as soon as I have time.

> No, that's why I now understand that it will not replace SG2/SG3 relay function. It's a complementary function to fine tune consumption but not changing HP behavior.

I think you are right. Wiping SG implementation will result in a feature removal, even if not used by the majority of ESPAltherma users. So I am in favour of keeping it, but you have the last word about this.

**Reply by xalex75** [reply]
I think that this could be very very interesting.
I started working on SG (just testing, heat pump is still in summer mode).
I created few "solar production levels" sensors:

1. up to 1000W
2. 1000W to 2000W
3. 2000W to 2500W
4. 2500W to 3000W
5. 3000W to 3500W
6. 3500W an Upper

I put these values into an input_select, so I can choose at which level SG will turn on (and off)
It works quite well, but solar production in winter is really depending on weather, here we have fog too, so it can vary very much.
Being able to set different consuption level can make my kind of management more efficient, because I can let SG turning on even if there is medium solar production, by limiting its power to the correct value.

![image](https://user-images.githubusercontent.com/52346449/197392783-41055f05-d4ca-4aa1-95f4-b2d687e9d770.png)

**Comment by alexdelli** [comment]
which pins use SG1 and SG2 relais?

**Comment by javimurcia** [comment]
Newer models that have the smart grid contacts integrated into it's main PCB (such as is the case of my EBBX116DF6V) has an additional input for the pulse output of a energy metter measuring the solar production.

In the "recommended on mode", the system will try to use power up to the limit set by the energy meter (the pulse rate).

My model also has the connector for the demand shapping pcb you are mentioning (as far as i know the pcb are just a bunch of optocouplers, so that extra pcb could be bypassed). If enabled, you can set a power (or current) limit for each of the inputs, and when active, the system will observe the most restrictive setting (ether comming from the demand shaping pcb or the smart grid interface).

Also, my unit has a "poor's man" smart grid, instead of giving it a pulse input for the power limit, yo can set said power limit in FW, so it will be the power limit of the "recommended on mode". This is what i'm trialing right now, while i wait for custom pcb that will generate said pulses for maximum PV divert.

**Reply by raomin** [reply]
Hi @javimurcia,

I'll have a look at the "recommended mode" and how we could integrate this. My understanding is that it would be the most flexible way of power limiting.

**Comment by javimurcia** [comment]
@raomin Great! i was hunting a photo of the backside of the demand shapping pcb, for reverse engeneering this function (it suks to spend more than $150 por a pcb with 4 optocouplers).

¿Does anyone know what is the name of the X85A connector, so we can build a nicer wiring harness?

**Reply by javimurcia** [reply]
I've figured what connector is the X85A using. It is a 6 pin JST-PA, wich has a 2.0mm pitch (so most probably, dupont jumpers won't fit) The exact part number is: PAP-06V-S don't forget to order the pins: SPHD-001T-P0.5
![image](https://user-images.githubusercontent.com/33467841/224303996-bd25da07-dc51-445a-b0a2-192eca5e701e.png)

**Watch out!** sometimes you will find Chinese HY 2.0 connectors being sold as JST-PH. **They won't fit!**
![image](https://user-images.githubusercontent.com/33467841/224302855-53a90c32-64c1-47c6-a4e3-56ab4f8c7a24.png)

I intend to use this feature so i can limit the heat pump power consumption when there is not solar production, so i'm not exceeding my available power when i also have other heavy consumers online (an ev charger for instance). I have implemented this function on https://github.com/raomin/ESPAltherma/pull/262

**Reply by maromme** [reply]
@javimurcia those PAP-06V-S are really difficult to get in a small number. where did you get those from?

**Reply by javimurcia** [reply]
> @javimurcia those PAP-06V-S are really difficult to get in a small number. where did you get those from?

I've got those from LCSC:
https://www.lcsc.com/product-detail/_JST-Sales-America-_C161867.html
https://www.lcsc.com/product-detail/_JST-Sales-America-_C246755.html
https://www.lcsc.com/product-detail/_JST-Sales-America-_C158044.html

**Reply by maromme** [reply]
thanks for your answer! it's a good hint!

**Reply by Dennis3321** [reply]
great find, put this information in the main documentation!@

**Reply by jf-wm** [reply]
Any update on the demand limiting function? Since the internal controller of my daikin unit sometimes doesn't work well, I thought maybe messing around with the damand limiting can mitigate these problems.

**Reply by gerione** [reply]
@Lectere: In Europe Conrad has them as well.

**Reply by DieterB61** [reply]
Conrad: https://github.com/raomin/ESPAltherma/issues/17#issuecomment-1783837967

**Reply by DieterB61** [reply]
@jf-wm: The limiting function works, you have to include the code shown here:
https://github.com/raomin/ESPAltherma/pull/262/files
@raomin: will you add this code to main version?

**Comment by raomin** [comment]
I guess in fact that the S4S / Pulse meter is what we should aim for, for consumption control. Setting SG1.SG2 on _Recommended ON_ then the heat pump limits it's consumption to the Pulse meter. Just need to know what the pulses are.
![image](https://user-images.githubusercontent.com/4460208/219984423-bb7964d3-7916-438a-b0d2-713b3d537848.png)

**Reply by javimurcia** [reply]
Some energy meters have a pulse output that works more or less like the light of a smart meter. Also there are sensors that can be placed over the smart meter light, which translates those light pulses into electric pulses 

Every time a fixed amount of energy flows (usually 1 watt hour*) thought the energy meter, it blinks.

Count the pulses and you have the energy recorded by the meter, **measure the time between pulses, and you know the power flowing thought the meter**.

What daikin is expecting is for you to have an energy meter measuring your pv production (in fact, that's how it's done in UK), and by feeding the pulses generated by that counter to the heat pump, it can know what is the power being generated by your pv array, and limit itself to that amount. 

If you have a more sophisticated setup, you might want to feed the heat pump something different, for instance I have information of my pv production, and my house's consumption (excluding the heat pump) so I will use the available pv power (pv production - house's consumption) instead of the pv production. 

It's something I was considering contributing to ESPAltherma , I think the most simple manner of implementing it is defining a topic where you can send the "available power", and ESPAltherma will convert it to a pulse rate on the sg pulse output pin. 

You just have to use an optocoupler, since a relay won't last long (1kW would output a 0.27Hz signal, or a pulse every 3.6  seconds, or 5.000 daily pulses, assuming 5 hours of light). 

That way, we can still use the m5stick c for this, by borrowing G0 from the main connector for the pulse output, and the 2 grove gpios for the smart grid signal, altought if you are interested in the "free mode" and "recommended on mode" that can be archived with a single gpio (the other signal is used by the forced on and forced off modes) 

The advantages of the X85A interface over the smart grid is that limitation is always observed, while the smart grid limitation is only followed when the smart grid is engaged in the recommended on mode. It's disadvantage is that you can only configure 4 discrete power levels while smart grid is more granular specifying the power limit.

So I'm going to play with both, and see what happens.

* if memory serves me right, for smart grids daikin only supports 100 pulses/kWh or 1.000 pulses/kWh

**Reply by maromme** [reply]
I found a setting in the reference guide about the pulse frequency. There is a setting:
![grafik](https://user-images.githubusercontent.com/60754017/230962389-831cf230-99b9-4f6f-8ca5-af771e86add8.png)

**Comment by maromme** [comment]
hi,
I read that discussion. It's just what I'll about implementing.  With the help of [Tardymo](https://github.com/raomin/ESPAltherma/issues/265) I got the espaltherma running smooth for a couple of weeks now (on a M5stick). Now it's about a bit more action there: I've got a EHSXB08P30EF (Altherma 3 R ECH2O) solar panels and a victron ESS and like to optimize the setup a bit.

First two things I'll implement are "on/off heating and cooling" and "SG1 and SG2". As that is documented it's now fuss there. I'm just waiting for the GPIO expander and the relaismodule to get started. 

How is developement for the implementation of the X85A interface and / or the pulse meter implementation going on? I would be happy if I could help, e.g. by testing.

Yours Markus

**Reply by maromme** [reply]
@raomin: I've got the hardware. tested SG1/SG2 and heating on/off. working fine as expected. how is development of the x85a going on? and especially is there any work on the port extender for the m5stick going on?

**Reply by rr0xx** [reply]
I bought the M5 4relay board to control the x85a but just found out these relays are controlled by the i2c protocol instead of the gpios. Does anyone have the experience how to integrate this in the ESPaltherma configuration?

**Comment by Kepro** [comment]
is there an option to use just a backup heater? without starting the heat pump? I want to minimize starts of the heat pump, I have plenty of energy for using just a backup heater (6 + 3kW) :-)

**Reply by pool1110** [reply]
I have a 9 kw heating rod.  In the settings I have limited the backup heater to 6KW and the third phase I have disconnected the cables and connected to a power regulator. 

I can heat with it up to 3 KW with PV electricity. As much as just photovoltaic power is produced.

So I save the switch-on cycles of the heat pump in the summer.

**Reply by Kepro** [reply]
yes! something like that, can you share some photos or schema? or both? 😅

**Reply by pool1110** [reply]
It is quite simple.

<img width="492" alt="image" src="https://github.com/raomin/ESPAltherma/assets/126029563/26af0e16-1573-4fbf-a782-4db27ecc2e71">

**Reply by pool1110** [reply]
Here is the power regulator:

https://de.aliexpress.com/item/32605706914.html?spm=a2g0o.new_account_index.0.0.66ee76c7i9LBdq&gatewayAdapt=glo2deu

**Reply by pool1110** [reply]
On the ESP8266 I have installed a PWM dimmer via ESPHome. 

With automations I always regulate the PWM signal so that the current surplus is 0 watts.

**Reply by Kepro** [reply]
okay, so you power directly backup heater on 1 phase, so how you enable water pump?

**Reply by pool1110** [reply]
I do not understand your question.

I heat the water with one phase from the 9KW heater. The Daikin Altherma then automatically turns on the water pump when it is needed.

**Reply by Kepro** [reply]
so it's done automatically? when there is power on phase Daikin will switch to DHW and start heat pump? 
I have 9KW which have built-in cable
![image](https://github.com/raomin/ESPAltherma/assets/1714370/2d8a02dc-a435-4970-b55a-d4a336e308a5)

![image](https://github.com/raomin/ESPAltherma/assets/1714370/d4c91009-9256-4f55-a42d-29e49f2c434d)

so can you point me to which contacts are you using? or show a picture maybe of your installation?

**Reply by pool1110** [reply]
I have the Daikin Altherma 3 R ECH2O. There ist the the connector XBUH1.

![IMG_5094](https://github.com/raomin/ESPAltherma/assets/126029563/acab9930-eb92-4828-94f0-3efbbec4c77f)

**Reply by pool1110** [reply]
And connected with the power regulator.
![IMG_5095](https://github.com/raomin/ESPAltherma/assets/126029563/1c2b2a42-b641-4934-a518-57306c659c3e)

**Reply by Kepro** [reply]
okay, that's cool, I need to check my altherma cuz I'm 99% sure I don't have XBUH1 :(

**Comment by maps2002** [comment]
Hi @pool1110 ,

So you are effectively regulating how much power is available to be used by the backup heater, right?
Would it be possible to share the core of the PWM/service regulation of your ESPHome file / YAML?

I would like to try something similar, and starting from your sucessfull basis would kick-start the work.

Thanks in advance!

> On the ESP8266 I have installed a PWM dimmer via ESPHome.
> 
> With automations I always regulate the PWM signal so that the current surplus is 0 watts.

**Comment by stomko11** [comment]
Revisiting this discussion. So I have indoor unit EHBH11CB9W and there is port X85A as stated in post nr.2 in this discussion. However I am not finding anything (any configuration) related to Smart Grid or power limitation. Is it that this model does not support it, or I need to "enable something somewhere" first and only then I will be able to control it? Or I am just unlucky and this feature is not for me? @raomin 
I would like to somehow integrate my HP and tell it to produce DHW while there is excessive energy coming from solar panels.
Ideas?

**Reply by stomko11** [reply]
I searched around a bit. I should have "Pwr. Consumpt. Control" item in menu A.6.(3), which i don't.
![image](https://github.com/raomin/ESPAltherma/assets/88718231/2110a875-9dfb-4239-83f1-37715fae9089)

I am thinking if this could be due to another setting in A.2.2.7 (Optional - adapter for power consumption limitation). At least that is what says in my native language

![image](https://github.com/raomin/ESPAltherma/assets/88718231/01c39df5-517e-4adc-ba0c-2f7f9e06ace4)

Would be nice if anyone could confirm if this could be it. What if I enable it. 
Also, I have no power meter, so my HP does not display anythin in regards to consumed energy, just produced. How would it limit the usage then?

**Reply by alexdelli** [reply]
Hi have the same HP model, but I don't understand how it works :(

**Reply by mx-vatte** [reply]
Although I've not tried it yet, there is no need for the additional PCB EKRP1AHT --> [https://github.com/raomin/ESPAltherma/discussions/170#discussioncomment-3909463](https://github.com/raomin/ESPAltherma/discussions/170#discussioncomment-3909463)

With ESPAltherma you are able to read/calculate the power consumption of your HP. With the power consumption control you could limit the HP consumption e.g. regarding your PV production/excess. To "tell it to produce DHW while there is excessive energy coming from solar panels" you should use the Smart Grid feature.

**Reply by mjonkman797** [reply]
I have the EHVX08, and there's no Smart Grid option available in the heat pump menu either. I purchased the BRP069A61 LAN adapter and hooked up two relays to the device running ESPaltherma. Only the "forced on" and "forced off" mode seem to actually do something. The "recommended on" mode doesn't do anything. A pulse meter should not be required since you can set a static power limitation in the configuration of the LAN adapter.
